File: L13ReceiptPrinter/Sources/PocketPrinter5890Kit/CodeBitmaps.swift
import Foundation
#if canImport(CoreImage)
import CoreImage
import CoreGraphics
#endif

/// Conversion des codes en bitmaps imprimables.
///
/// Les codes-barres lineaires sont generes en Swift pur (`Barcode1D`), donc
/// portables partout. Les QR codes s'appuient sur CoreImage: ecrire un
/// encodeur QR conforme (masques, evaluation de penalite, format BCH) est
/// une source d'erreurs silencieuses, et un QR faux ne se voit qu'au
/// scanner. Le moteur est isole ici pour pouvoir etre remplace sans toucher
/// au reste de la librairie.
///
/// Dans tous les cas les modules sont ecrits directement dans le bitmap,
/// sans interpolation: un code redimensionne en douceur devient illisible.
public enum CodeBitmaps {

    /// QR code centre sur la largeur d'impression.
    ///
    /// - Parameters:
    ///   - content: donnees a encoder.
    ///   - correction: niveau de correction d'erreur.
    ///   - moduleSize: taille d'un module en pixels. Calculee automatiquement
    ///     pour occuper environ deux tiers de la largeur si elle vaut nil.
    ///   - quietZone: marge blanche autour du code, en modules. La norme en
    ///     exige 4; en dessous, beaucoup de lecteurs echouent.
    public static func qrCode(
        _ content: String,
        correction: QRCorrectionLevel = .medium,
        moduleSize: Int? = nil,
        quietZone: Int = 4,
        printWidth: Int = MonochromeBitmap.nativeWidth
    ) throws -> MonochromeBitmap {
        guard !content.isEmpty else { throw CodeError.emptyContent }
        let matrix = try qrMatrix(content, correction: correction)
        let totalModules = matrix.size + quietZone * 2

        // Taille de module par defaut: le code occupe environ deux tiers de
        // la largeur, sans jamais la depasser.
        let scale = moduleSize ?? max(1, (printWidth * 2 / 3) / totalModules)
        let codeSize = min(totalModules * scale, printWidth)

        let widthBytes = (printWidth + 7) / 8
        var bytes = [UInt8](repeating: 0, count: widthBytes * codeSize)

        let originX = (printWidth - codeSize) / 2
        for y in 0..<codeSize {
            let moduleY = y / scale - quietZone
            guard moduleY >= 0, moduleY < matrix.size else { continue }
            for x in 0..<codeSize {
                let moduleX = x / scale - quietZone
                guard moduleX >= 0, moduleX < matrix.size else { continue }
                guard matrix[moduleX, moduleY] else { continue }
                let pixelX = originX + x
                bytes[y * widthBytes + pixelX / 8] |= UInt8(0x80 >> (pixelX % 8))
            }
        }
        return try MonochromeBitmap(width: printWidth, height: codeSize, bytes: bytes)
    }

    public enum QRCorrectionLevel: String, CaseIterable, Identifiable, Sendable {
        case low = "L"
        case medium = "M"
        case quartile = "Q"
        case high = "H"

        public var id: String { rawValue }

        public var title: String {
            switch self {
            case .low: return "Faible (7 %)"
            case .medium: return "Moyen (15 %)"
            case .quartile: return "Bon (25 %)"
            case .high: return "Maximal (30 %)"
            }
        }
    }

    public enum CodeError: Error, LocalizedError {
        case emptyContent
        case generationFailed
        case unavailable

        public var errorDescription: String? {
            switch self {
            case .emptyContent: return "Contenu vide"
            case .generationFailed: return "Generation du QR code impossible"
            case .unavailable: return "Generateur de QR code indisponible sur cette plateforme"
            }
        }
    }

    /// Matrice de modules d'un QR code.
    public struct QRMatrix {
        public let size: Int
        public let modules: [Bool]

        public subscript(x: Int, y: Int) -> Bool {
            modules[y * size + x]
        }
    }

    /// Produit la matrice de modules, sans mise a l'echelle.
    ///
    /// CoreImage rend le code avec une quiet zone d'un module: elle est
    /// retiree ici pour que l'appelant maitrise sa propre marge.
    public static func qrMatrix(
        _ content: String,
        correction: QRCorrectionLevel = .medium
    ) throws -> QRMatrix {
        #if canImport(CoreImage)
        guard let filter = CIFilter(name: "CIQRCodeGenerator") else {
            throw CodeError.unavailable
        }
        filter.setValue(Data(content.utf8), forKey: "inputMessage")
        filter.setValue(correction.rawValue, forKey: "inputCorrectionLevel")
        guard let output = filter.outputImage else { throw CodeError.generationFailed }

        let width = Int(output.extent.width)
        let height = Int(output.extent.height)
        guard width > 2, height > 2 else { throw CodeError.generationFailed }

        let context = CIContext(options: [.useSoftwareRenderer: true])
        guard let cgImage = context.createCGImage(output, from: output.extent) else {
            throw CodeError.generationFailed
        }

        var pixels = [UInt8](repeating: 0, count: width * height * 4)
        pixels.withUnsafeMutableBytes { buffer in
            guard let ctx = CGContext(
                data: buffer.baseAddress,
                width: width,
                height: height,
                bitsPerComponent: 8,
                bytesPerRow: width * 4,
                space: CGColorSpaceCreateDeviceRGB(),
                bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
            ) else { return }
            ctx.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
        }

        // CoreImage ajoute une bordure d'un module: on la retire.
        let border = 1
        let size = width - border * 2
        var modules = [Bool](repeating: false, count: size * size)
        for y in 0..<size {
            for x in 0..<size {
                let offset = ((y + border) * width + (x + border)) * 4
                modules[y * size + x] = pixels[offset] < 128
            }
        }
        return QRMatrix(size: size, modules: modules)
        #else
        throw CodeError.unavailable
        #endif
    }

    /// Code-barres lineaire centre sur la largeur d'impression.
    ///
    /// - Parameters:
    ///   - moduleWidth: largeur d'une barre elementaire. Calculee pour
    ///     remplir la largeur disponible si elle vaut nil.
    ///   - quietZone: marge blanche laterale en pixels.
    public static func barcode(
        _ content: String,
        symbology: Barcode1D.Symbology = .code128,
        height: Int = 80,
        moduleWidth: Int? = nil,
        quietZone: Int = 20,
        printWidth: Int = MonochromeBitmap.nativeWidth
    ) throws -> MonochromeBitmap {
        let pattern = try Barcode1D.pattern(for: content, symbology: symbology)
        let available = max(1, printWidth - quietZone * 2)
        let scale = moduleWidth ?? max(1, available / pattern.count)
        let codeWidth = min(pattern.count * scale, printWidth)

        let height = max(8, height)
        let widthBytes = (printWidth + 7) / 8
        var bytes = [UInt8](repeating: 0, count: widthBytes * height)

        let originX = (printWidth - codeWidth) / 2
        for x in 0..<codeWidth {
            let index = x / scale
            guard index < pattern.count, pattern[index] else { continue }
            let pixelX = originX + x
            for y in 0..<height {
                bytes[y * widthBytes + pixelX / 8] |= UInt8(0x80 >> (pixelX % 8))
            }
        }
        return try MonochromeBitmap(width: printWidth, height: height, bytes: bytes)
    }
}

public extension PrintElement {
    /// QR code genere en Swift pur, portable et sans dependance systeme.
    ///
    /// A preferer a `.qrCode(_:moduleSize:)`, qui repose sur la commande
    /// native `GS ( k` que ce firmware n'implemente pas.
    static func qr(
        _ content: String,
        correction: CodeBitmaps.QRCorrectionLevel = .medium,
        moduleSize: Int? = nil,
        printWidth: Int = MonochromeBitmap.nativeWidth
    ) throws -> PrintElement {
        .image(try CodeBitmaps.qrCode(
            content,
            correction: correction,
            moduleSize: moduleSize,
            printWidth: printWidth
        ))
    }

    /// Code-barres genere en Swift pur.
    static func code(
        _ content: String,
        symbology: Barcode1D.Symbology = .code128,
        height: Int = 80,
        printWidth: Int = MonochromeBitmap.nativeWidth
    ) throws -> PrintElement {
        .image(try CodeBitmaps.barcode(
            content,
            symbology: symbology,
            height: height,
            printWidth: printWidth
        ))
    }
}

File: android/kit/src/main/kotlin/fr/pocketprinter5890/kit/CodeBitmaps.kt
package fr.pocketprinter5890.kit

import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.qrcode.QRCodeWriter
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel

/** Erreurs de generation d'un code en bitmap. */
sealed class CodeException(message: String) : Exception(message) {
    object EmptyContent : CodeException("Contenu vide") {
        private fun readResolve(): Any = EmptyContent
    }

    object GenerationFailed : CodeException("Generation du code impossible") {
        private fun readResolve(): Any = GenerationFailed
    }
}

/**
 * Conversion des codes en bitmaps imprimables.
 *
 * Les codes-barres lineaires sont generes en Kotlin pur ([Barcode1D]), donc
 * portables partout. Les QR codes s'appuient sur ZXing: ecrire un encodeur QR
 * conforme (masques, evaluation de penalite, format BCH) est une source
 * d'erreurs silencieuses, et un QR faux ne se voit qu'au scanner. C'est aussi
 * la bibliotheque qu'utilise l'application officielle. Le moteur est isole ici
 * pour pouvoir etre remplace sans toucher au reste de la librairie.
 *
 * Dans tous les cas les modules sont ecrits directement dans le bitmap, sans
 * interpolation: un code redimensionne en douceur devient illisible.
 */
object CodeBitmaps {

    enum class QRCorrectionLevel(val title: String) {
        LOW("Faible (7 %)"),
        MEDIUM("Moyen (15 %)"),
        QUARTILE("Bon (25 %)"),
        HIGH("Maximal (30 %)");

        internal val zxing: ErrorCorrectionLevel
            get() = when (this) {
                LOW -> ErrorCorrectionLevel.L
                MEDIUM -> ErrorCorrectionLevel.M
                QUARTILE -> ErrorCorrectionLevel.Q
                HIGH -> ErrorCorrectionLevel.H
            }
    }

    /** Matrice de modules d'un QR code, sans zone de silence. */
    class QRMatrix(val size: Int, val modules: BooleanArray) {
        operator fun get(x: Int, y: Int): Boolean = modules[y * size + x]
    }

    /**
     * Genere la matrice brute d'un QR code, sans mise a l'echelle.
     *
     * ZXing ajoute une zone de silence par defaut; on la desactive ici pour
     * maitriser nous-memes la marge dans [qrCode].
     */
    @Throws(CodeException::class)
    fun qrMatrix(
        content: String,
        correction: QRCorrectionLevel = QRCorrectionLevel.MEDIUM
    ): QRMatrix {
        if (content.isEmpty()) throw CodeException.EmptyContent
        val hints = mapOf(
            EncodeHintType.ERROR_CORRECTION to correction.zxing,
            EncodeHintType.CHARACTER_SET to "UTF-8",
            // La zone de silence est ajoutee par qrCode(), pas par ZXing.
            EncodeHintType.MARGIN to 0
        )
        val matrix = runCatching {
            QRCodeWriter().encode(content, BarcodeFormat.QR_CODE, 0, 0, hints)
        }.getOrElse { throw CodeException.GenerationFailed }

        val size = matrix.width
        if (size <= 0 || matrix.height != size) throw CodeException.GenerationFailed

        val modules = BooleanArray(size * size)
        for (y in 0 until size) {
            for (x in 0 until size) {
                modules[y * size + x] = matrix.get(x, y)
            }
        }
        return QRMatrix(size, modules)
    }

    /**
     * QR code centre sur la largeur d'impression.
     *
     * @param moduleSize taille d'un module en pixels. Calculee automatiquement
     *   pour occuper environ deux tiers de la largeur si elle vaut null.
     * @param quietZone marge blanche autour du code, en modules. La norme en
     *   exige 4; en dessous, beaucoup de lecteurs echouent.
     */
    @Throws(CodeException::class, RasterException::class)
    fun qrCode(
        content: String,
        correction: QRCorrectionLevel = QRCorrectionLevel.MEDIUM,
        moduleSize: Int? = null,
        quietZone: Int = 4,
        printWidth: Int = MonochromeBitmap.nativeWidth
    ): MonochromeBitmap {
        if (content.isEmpty()) throw CodeException.EmptyContent
        val matrix = qrMatrix(content, correction)
        val totalModules = matrix.size + quietZone * 2

        // Taille de module par defaut: le code occupe environ deux tiers de la
        // largeur, sans jamais la depasser.
        val scale = moduleSize ?: maxOf(1, (printWidth * 2 / 3) / totalModules)
        val codeSize = minOf(totalModules * scale, printWidth)

        val widthBytes = (printWidth + 7) / 8
        val bytes = ByteArray(widthBytes * codeSize)

        val originX = (printWidth - codeSize) / 2
        for (y in 0 until codeSize) {
            val moduleY = y / scale - quietZone
            if (moduleY < 0 || moduleY >= matrix.size) continue
            for (x in 0 until codeSize) {
                val moduleX = x / scale - quietZone
                if (moduleX < 0 || moduleX >= matrix.size) continue
                if (!matrix[moduleX, moduleY]) continue
                val pixelX = originX + x
                val index = y * widthBytes + pixelX / 8
                bytes[index] = (bytes[index].toInt() or (0x80 shr (pixelX % 8))).toByte()
            }
        }
        return MonochromeBitmap(printWidth, codeSize, bytes)
    }

    /**
     * Code-barres lineaire centre sur la largeur d'impression.
     *
     * @param moduleWidth largeur d'une barre elementaire. Calculee pour remplir
     *   la largeur disponible si elle vaut null.
     * @param quietZone marge blanche laterale en pixels.
     */
    @Throws(BarcodeException::class, RasterException::class)
    fun barcode(
        content: String,
        symbology: Barcode1D.Symbology = Barcode1D.Symbology.CODE128,
        height: Int = 80,
        moduleWidth: Int? = null,
        quietZone: Int = 20,
        printWidth: Int = MonochromeBitmap.nativeWidth
    ): MonochromeBitmap {
        val pattern = Barcode1D.pattern(content, symbology)
        val available = maxOf(1, printWidth - quietZone * 2)
        val scale = moduleWidth ?: maxOf(1, available / pattern.size)
        val codeWidth = minOf(pattern.size * scale, printWidth)

        val codeHeight = maxOf(8, height)
        val widthBytes = (printWidth + 7) / 8
        val bytes = ByteArray(widthBytes * codeHeight)

        val originX = (printWidth - codeWidth) / 2
        for (x in 0 until codeWidth) {
            val index = x / scale
            if (index >= pattern.size || !pattern[index]) continue
            val pixelX = originX + x
            for (y in 0 until codeHeight) {
                val offset = y * widthBytes + pixelX / 8
                bytes[offset] = (bytes[offset].toInt() or (0x80 shr (pixelX % 8))).toByte()
            }
        }
        return MonochromeBitmap(printWidth, codeHeight, bytes)
    }
}

/**
 * QR code genere en bitmap, portable et sans commande native.
 *
 * A preferer a [PrintElement.QrCode], qui repose sur la commande native
 * `GS ( k` que ce firmware n'implemente pas.
 */
@Throws(CodeException::class, RasterException::class)
fun PrintElement.Companion.qr(
    content: String,
    correction: CodeBitmaps.QRCorrectionLevel = CodeBitmaps.QRCorrectionLevel.MEDIUM,
    moduleSize: Int? = null,
    printWidth: Int = MonochromeBitmap.nativeWidth
): PrintElement = PrintElement.Image(
    CodeBitmaps.qrCode(content, correction, moduleSize, printWidth = printWidth)
)

/** Code-barres genere en bitmap, seule voie fiable sur ce firmware. */
@Throws(BarcodeException::class, RasterException::class)
fun PrintElement.Companion.code(
    content: String,
    symbology: Barcode1D.Symbology = Barcode1D.Symbology.CODE128,
    height: Int = 80,
    printWidth: Int = MonochromeBitmap.nativeWidth
): PrintElement = PrintElement.Image(
    CodeBitmaps.barcode(content, symbology, height, printWidth = printWidth)
)
